### PR TITLE
feat: Dataset table view

### LINKED ISSUE #366: Dataset Table View
## Problem

There is current no way for our users to view a dataset as a table, only in an image grid. This is required for visualising both imaging and/or tabular datasets.

Relates to https://github.com/gliff-ai/roadmap/issues/29 
Relates to https://github.com/gliff-ai/roadmap/issues/30

## Solution

**CURATE** requires a reshuffle to better organise the UI elements. The control panel on the left is now organised:

1. The `S` `M` and `L` icons offer the ability to change image size. The `image` and `table` icons offer the ability to change dataset view. These are grouped together and next to each other as a way to better connect these.
2. The `user` `labels` `plugins` icon is the tools for the dataset. While these require rethinking as the similar functionality of `user` and `labels` also exists elsewhere. These are grouped together for now.
3. The `sort` and `search` are the same.
4. The `annotation progress` is awaiting implementation.
5. The `upload` and `download` are grouped together at the bottom. These are more external tools.

Curate – Dataset Grid  Layout

The two `image` and `grid` view icons should have these tooltips - shortcut assignment pending(?).

Curate – Dataset Grid  Icon Hover State – Dataset Grid Curate – Dataset Grid  Icon Hover State – Dataset Table

The `table` view is very similar to **MANAGE**. Some slight tweaks which are working towards revamping how we handle tables across the platform.

The design assumes a dataset might have a lot of metadata resulting in a lot of columns and row. To handle this the table element should be scrollable left and right and up and up for instances where this is the case. The UI element itself does not run off the page, i.e. how GitHub handles text boxes - scrollable but fixed size. The column headers and green top-bar should remain static, this gives the user a reference to what the data is if they scroll far down.

Curate – Dataset Table  Layout

The settings icon in the right of the green top-bar allows users to customise the view of the table. Separated into `visible fields` and `hidden fields`. If a user clicks a `hidden field` it should move up and vice versa. This provides users more information of their interactions with the dialogue. 

Curate – Dataset Table  Dialogue – Edit Table Fields

## Considered Alternatives

I considered wether the green top-bar was needed and how to create that table hierarchy (with background colour or the addition of dividing lines). For now, I kept the styling similar but this may be revisited during the polishing of the UI design system and what may be considered the best direction for platform usability.

Curate – Dataset Table – v2 1

### PR DESCRIPTION
**Pull request by ChrisBaidoo**
## Description

Import Datagrid component from STYLE to view a dataset as a table.

closes #366

## Checklist:

Put an `x` in the boxes that apply to this pull request (you can also fill these out after opening the pull request). If you're unsure about any of these, don't hesitate to leave a comment on this pull request!

- [x] I have read the gliff.ai Contribution Guide.
- [x] I have requested to **pull a branch** and not from main.
- [x] I have checked all commit message styles match the requested structure.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my own code.
- [ ] I have assigned **3 or less** reviewers.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] My changes generate no new warnings.
- [ ] I have made corresponding changes to the documentation.
- [ ] New database changes have been committed.
- [ ] If appropriate, I have bumped any version numbers.

### CODE REVIEW

File: src/components/Tooltips.ts [hunk 1]
```diff
@@ -9,6 +9,25 @@ interface ThumbnailSizes {
   id?: string;
 }
 
+interface DatasetType {
```
Comment (cooper667) [inline src/components/Tooltips.ts:12]: Maybe `DatasetViewType` as the name?

Comment (cooper667) [reply src/components/Tooltips.ts:12]: Also doesn't need exporting?

File: src/ui.tsx [hunk 2]
```diff
@@ -750,6 +714,237 @@ class UserInterface extends Component<Props, State> {
...
+        headerName: "Labels",
+        field: "labels",
+        width: 250,
+        editable: false,
+        renderCell: (params: GridRenderCellParams<any, any, any>) => {
+          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+          const imageLabels: string[] = params?.row?.imageLabels || [];
+          const chipsLimit = 2;
+          const chipsContent: string[] = imageLabels.slice(0, chipsLimit);
+          const moreLabelsCount = imageLabels.length - 2;
```
Comment (cooper667) [inline src/ui.tsx:746]:
```suggestion
          const moreLabelsCount = imageLabels.length - chipsLimit;
```

File: src/ui.tsx [hunk 3]
```diff
@@ -750,6 +714,237 @@ class UserInterface extends Component<Props, State> {
...
+        field: "assignees",
+        width: 150,
+        editable: false,
+      },
+      {
+        headerName: "Labels",
+        field: "labels",
+        width: 250,
+        editable: false,
+        renderCell: (params: GridRenderCellParams<any, any, any>) => {
```
Comment (cooper667) [inline src/ui.tsx:741]:
```suggestion
renderCell: (params: GridRenderCellParams<unknown, MetaItem, unknown>) => {
```

should remove the need to ignore ES Lint below

File: src/ui.tsx [hunk 4]
```diff
@@ -750,6 +714,237 @@ class UserInterface extends Component<Props, State> {
...
+
+    const allCols = [...defaultColumns, ...Object.values(colsObj)];
+
+    const tableView = (
+      <Box sx={{ width: "100%", marginTop: "14px", height: "700px" }}>
+        <DataGrid
+          title="Dataset Details"
+          columns={allCols}
+          rows={this.state.metadata}
+          sx={{ height: "600px" }}
```
Comment (cooper667) [inline src/ui.tsx:808]: This needs to fill the page rather than be fixed

**Review by cooper667 — COMMENTED** [review]
To fix the build errors I _think_ we need to bump Material UI, which creates other problems...I'm going to try that in a different branch!

File: src/ui.tsx (hunk 3 above)
Comment (cooper667) [reply src/ui.tsx:741]: and also you might not need to cast to `string[]` on the line below

**Review by joshuajames-smith — CHANGES_REQUESTED** [review]
Screenshot 2022-05-26 at 14 17 53

- [ ] Can the datagrid align to the bottom? Similar to how in MANAGE the tables align at the bottom!
- [ ] Can we set some rules to have space between the columns? There is two columns in staging that are next to each other and would benefit from visual space between them.
- [x] We use pagination but only have 1 arrow - assuming this is done automatically? Can we either abandoned pagination for a full scrollable list or increase number of results shown at once (maybe 60?) and add in full pagination arrows?

**Comment by cooper667** [comment]
> We use pagination but only have 1 arrow - assuming this is done automatically? Can we either abandoned pagination for a full scrollable list or increase number of results shown at once (maybe 60?) and add in full pagination arrows?

Regarding this, should we just NOT have pagination as we have always loaded the full dataset (_i think_) so it's not like we need to implement infinite scroll or anything, we have it all and the pagination is currently artificial? Then maybe add pagination in the future when we actually lazy load the data?

(one for @joshuajames-smith and/or @ChasNelson1990  and/or @ChrisBaidoo )

**Review by cooper667 — DISMISSED** [review]

**Comment by joshuajames-smith** [comment]
> > We use pagination but only have 1 arrow - assuming this is done automatically? Can we either abandoned pagination for a full scrollable list or increase number of results shown at once (maybe 60?) and add in full pagination arrows?
> 
> Regarding this, should we just NOT have pagination as we have always loaded the full dataset (_i think_) so it's not like we need to implement infinite scroll or anything, we have it all and the pagination is currently artificial? Then maybe add pagination in the future when we actually lazy load the data?
> 
> (one for @joshuajames-smith and/or @ChasNelson1990 and/or @ChrisBaidoo )

Last spoke to @ChasNelson1990 I believe something was mentioned how pagination is good for E2E? But from my POV im happy with full data load with scroll - this is supported by the filter/search system anyway!

**Review by joshuajames-smith — COMMENTED** [review]
@ChrisBaidoo is it possible to load all the data at once and remove pagination?

**Review by cooper667 — APPROVED** [review]

**Review by ChasNelson1990 — APPROVED** [review]
Great work, looks good to me!

**Review by joshuajames-smith — APPROVED** [review]

**Pull request merged**